scenes/globals/pause.gd

```diff
@@ -0,0 +1,114 @@
+extends Node
+
+enum System {
+	PLAYER_INPUT,
+	GAME
+	# For now is just these two but in the future we could have: Physics, Enemies, AI, etc
+}
+
+# Private vars
+var _pause_requests: Array[PauseRequest] = []
+
+# This should be an array of systems but it throws
+# a runtime error when using the filter function, so for now just "Array"
+var _paused_systems: Array = []
+
+
+# Public interface
+func pause_system(system: System, node: Node) -> void:
+	_add_request(system, node)
+
+
+func unpause_system(system: System, node: Node) -> void:
+	_clear_request(system, node)
+
+	_refresh_systems()
+
+
+func is_paused(system: System) -> bool:
+	return system in paused_systems()
+
+
+func paused_systems() -> Array:
+	return _paused_systems
+
+
+func unpaused_systems() -> Array:
+	return System.values().filter(func(system: System) -> bool: return !is_paused(system))
+
+
+# Private methods
+func _find_request(a_system: System, a_node: Node) -> PauseRequest:
+	for req: PauseRequest in _pause_requests:
+		if req.node == a_node and req.system == a_system:
+			return req
+
+	return null
+
+
+func _exists_request_for_system(a_system: System) -> bool:
+	return _pause_requests.any(func(req: PauseRequest) -> bool: return req.system == a_system)
+
+
+func _add_request(system: System, a_node: Node) -> void:
+	var req: PauseRequest = _find_request(system, a_node)
+
+	if !req:
+		_pause_requests.push_back(PauseRequest.new(system, a_node))
+		a_node.tree_exited.connect(self._remove_pauses_of_node.bind(a_node), CONNECT_ONE_SHOT)
+
+		_refresh_systems()
+	else:
+		print(
+			(
+				"Node %s requested pause for system %s but already exists"
+				% [a_node.name, System.find_key(system)]
+			)
+		)
+
+
+func _clear_request(system: System, a_node: Node) -> void:
+	var req: PauseRequest = _find_request(system, a_node)
+
+	if req:
+		_pause_requests.erase(req)
+		if is_instance_valid(req.node):
+			req.node.tree_exited.disconnect(self._remove_pauses_of_node.bind(system, a_node))
+
+		_refresh_systems()
+
+
+func _refresh_systems() -> void:
+	_refresh_paused_systems()
+
+	if not is_inside_tree():
+		# This can happen when the game is quitting
+		return
+
+	if is_paused(System.GAME):
+		get_tree().paused = true
+	else:
+		get_tree().paused = false
+
+
+func _refresh_paused_systems() -> void:
+	_paused_systems = System.values().filter(
+		func(system: System) -> bool: return _exists_request_for_system(system)
+	)
+
+
+func _remove_pauses_of_node(removed_node: Node) -> void:
+	for req: PauseRequest in _pause_requests.duplicate():
+		if req.node == removed_node:
+			_pause_requests.erase(req)
+
+	_refresh_systems()
+
+
+class PauseRequest:
```

Comment:
It took me some time to find what was `PauseRequest` until I found it at the bottom of the script. Is the linter forcing this? Otherwise can it be moved above all the `func` declarations?

Reply:
The linter forces me to have it below :(